### PR TITLE
sql: make ctxWithStacktrace compatible with gRPC

### LINKED ISSUE #96457: sql: make ctxWithStacktrace compatible with gRPC
`type ctxWithStacktrace struct` was added here https://github.com/cockroachdb/cockroach/pull/95797 to add a stacktrace (via `errors.WithStack(...)` when `Context.Err()` is called inside `contextutil.RunWithTimeout(...)`.

This ended up causing an issue if gRPC calls `ctxWithStacktrace.Err()` because it is expecting `context.DeadlineExceeded` to not be wrapped with `errors.WithStack`- the wrong status code will be returned in the error returned here https://github.com/grpc/grpc-go/blob/v1.46.0/rpc_util.go#L833.

To resolve this, store the wrapped exception separately inside `ctxWithStacktrace` when `ctxWithStacktrace.Err()` is called and set this to `TimeoutError.cause` inside `contextutil.RunWithTimeout(...)`.

See https://github.com/cockroachdb/cockroach/pull/96289

Jira issue: CRDB-24132

### PR DESCRIPTION
**Pull request by ecwall**
Fixes #96457

`type ctxWithStacktrace struct` was added here #95797 to add a stacktrace
(via `errors.WithStack(...)` when `Context.Err()` is called inside
`contextutil.RunWithTimeout(...)`.

This ended up causing an issue if gRPC calls `ctxWithStacktrace.Err()`
because it is expecting `context.DeadlineExceeded` to not be wrapped with
`errors.WithStack` -  the wrong status code will be returned in the error
returned here https://github.com/grpc/grpc-go/blob/v1.46.0/rpc_util.go#L833.

To resolve this, store the wrapped exception separately inside
`ctxWithStacktrace` when `ctxWithStacktrace.Err()` is called and set this to
`TimeoutError.cause` inside `contextutil.RunWithTimeout(...)`.

Release note: None

### CODE REVIEW

File: pkg/util/contextutil/context.go [hunk 1]
```diff
@@ -79,16 +80,37 @@ func wrap(ctx context.Context, cancel context.CancelFunc) (context.Context, cont
...
 
 // Err implements the context.Context interface.
-func (ctx *ctxWithStacktrace) Err() error {
-	return errors.WithStack(ctx.Context.Err())
+func (ctx *contextWithStacktrace) Err() error {
+	err := ctx.Context.Err()
+	if err == nil {
+		return nil
+	}
+	ctx.mu.Lock()
```
Comment (rafiss) [inline pkg/util/contextutil/context.go:107]: do you think this would be a good place to use syncutil.Once https://pkg.go.dev/go4.org/syncutil#Once.Do

Comment (ecwall) [reply pkg/util/contextutil/context.go:107]: `Once` would work if `errWithStacktrace` were only being accessed from inside `Do`, but it is also accessed from `errWithStacktrace` so a mutex is needed that guards both.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, and @rafiss)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNGwKkBcQuK24kdSm7):*
I know I've suggested this, but I believe it doesn't work when the context expiration is detected by calling `Err` on a child ctx created by `WithCancel` (and maybe also by `WithDeadline/WithTimeout`, I'm not sure) - which I think is the common case when the deadline comes from an rpc call or a sql query or most things, really.
If that's true, I'd reconsider whether having this hack is worth it. Is it ever going to help?
Or, to phrase it as a question - have you verified that this actually shows an interesting stack in at least some cases for the particular deadline that prompted this patch?
___
*[`pkg/util/contextutil/context.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNDsN0-E0G2i_2g7nv:-NNNDsN0-E0G2i_2g7nw:b-q2lgbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/contextutil/context.go#L91)):*
> ```Go
> 	mu struct {
> 		syncutil.Mutex
> 		errWithStacktrace error
> ```

maybe `capturedErr`
___
*[`pkg/util/contextutil/context.go` line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNDA9g5f6Zkm4WAaNb:-NNNDA9g5f6Zkm4WAaNc:b-u7tilz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/contextutil/context.go#L95)):*
> ```Go
> }
> 
> func (ctx *contextWithStacktrace) ErrWithStacktrace() error {
> ```

Let's find a more suggestive name for this. Maybe `CapturedErr`?
And put a comment saying something about how it indicates the first detector of the expiration (kinda, since it doesn't deal with `Done`).
And also say something about how it doesn't work if there's a `context.WithCancel` under it.

Also, don't export it if the type is not exported.
___
*[`pkg/util/contextutil/context.go` line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNEi345dPmwiBy5LWb:-NNNEi345dPmwiBy5LWc:b7nhs6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/contextutil/context.go#L95)):*
> ```Go
> }
> 
> func (ctx *contextWithStacktrace) ErrWithStacktrace() error {
> ```

As written, this can return nil even for an expired ctx. End even for a context that was detected to have expired (through calling `Err` on a child produced by `context.WithCancel`. Are you sure that's what you want?
The one caller you have does use `Combine()`, which deals with a nil, but still.
Consider calling `Err()`.
___
*[`pkg/util/tracing/grpcinterceptor/grpc_interceptor_test.go` line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNIPfx2UpPhsI5RG-9:-NNNIPfx2UpPhsI5RG-A:b1s30jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/tracing/grpcinterceptor/grpc_interceptor_test.go#L134)):*
> ```Go
> 		},
> 		{
> 			name:     "UnaryUnary_ContextTimeout_DeadlineExceeded",
> ```

I don't think the new tests are worth it. I think all we need to test is that our ctx implementation respects the contract. I think these "integration tests" will no t make much sense to any reader.

File: pkg/util/contextutil/context.go (hunk 1 above)
Comment (rafiss) [reply pkg/util/contextutil/context.go:107]: ah ok, makes sense

**Review by rafiss — APPROVED** [review]

**Comment by rafiss** [comment]
oh oops, missed Andrei's comments. I'll leave the review to him

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @otan)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNGwKkBcQuK24kdSm7):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I know I've suggested this, but I believe it doesn't work when the context expiration is detected by calling `Err` on a child ctx created by `WithCancel` (and maybe also by `WithDeadline/WithTimeout`, I'm not sure) - which I think is the common case when the deadline comes from an rpc call or a sql query or most things, really.
If that's true, I'd reconsider whether having this hack is worth it. Is it ever going to help?
Or, to phrase it as a question - have you verified that this actually shows an interesting stack in at least some cases for the particular deadline that prompted this patch?
</blockquote></details>

I set the auth timeout to `1ns` and the returned error generates a stacktrace like this which is what we want:
```
operation "get-user-timeout" timed out after 0s (given timeout 1ns): internal error while retrieving user account: context deadline exceeded
(1) operation "get-user-timeout" timed out after 0s (given timeout 1ns)
Wraps: (2) secondary error attachment
  | captured context error: context deadline exceeded
  | (1) attached stack trace
  |   -- stack trace:
  |   | github.com/cockroachdb/cockroach/pkg/util/contextutil.(*contextWithStacktrace).Err
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:116
  |   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:924
  |   | github.com/cockroachdb/cockroach/pkg/kv.runTxn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:959
  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:922
  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:897
  |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:1471
  |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).DescsTxn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:1393
  |   | github.com/cockroachdb/cockroach/pkg/sql/sessioninit.(*Cache).GetAuthInfo
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/sessioninit/cache.go:124
  |   | github.com/cockroachdb/cockroach/pkg/sql.retrieveSessionInitInfoWithCache.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:210
  |   | github.com/cockroachdb/cockroach/pkg/sql.retrieveSessionInitInfoWithCache
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:234
  |   | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1.1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:98
  |   | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:131
  |   | github.com/cockroachdb/cockroach/pkg/sql.getUserInfoRunFn.func2
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:190
  |   | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:97
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.authAutoSelectPasswordProtocol.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_methods.go:509
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*AuthBehaviors).Authenticate
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_behaviors.go:56
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).handleAuthentication
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth.go:178
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:693
  |   | runtime.goexit
  |   | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1594
  | Wraps: (2) captured context error
  | Wraps: (3) context deadline exceeded
  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) context.deadlineExceededError
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.retrieveSessionInitInfoWithCache
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:237
  | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1.1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:98
  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:131
  | github.com/cockroachdb/cockroach/pkg/sql.getUserInfoRunFn.func2
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:190
  | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:97
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.authAutoSelectPasswordProtocol.func1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_methods.go:509
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*AuthBehaviors).Authenticate
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_behaviors.go:56
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).handleAuthentication
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth.go:178
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:693
  | runtime.goexit
  | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1594
Wraps: (4) internal error while retrieving user account
Wraps: (5) context deadline exceeded
  | -- cause hidden behind barrier
  | context deadline exceeded
  | (1) context deadline exceeded
  | Error types: (1) context.deadlineExceededError
Error types: (1) *contextutil.TimeoutError (2) *secondary.withSecondaryError (3) *withstack.withStack (4) *errutil.withPrefix (5) *barriers.barrierErr
```

But you are right, it is possible to wrap `contextWithStacktrace` in another `Context` and call `Err()` on it instead which will not generate a useful stacktrace (I added a test case proving this).
___
*[`pkg/util/contextutil/context.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNDsN0-E0G2i_2g7nv:-NNNPpEhCVWPKZ_iS6uN:bt1b19p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/contextutil/context.go#L91)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe `capturedErr`
</blockquote></details>

Renamed.
___
*[`pkg/util/contextutil/context.go` line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNDA9g5f6Zkm4WAaNb:-NNNaglNAmPDQCSyIdq7:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/contextutil/context.go#L95)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Let's find a more suggestive name for this. Maybe `CapturedErr`?
And put a comment saying something about how it indicates the first detector of the expiration (kinda, since it doesn't deal with `Done`).
And also say something about how it doesn't work if there's a `context.WithCancel` under it.

Also, don't export it if the type is not exported.
</blockquote></details>

Updated.
___
*[`pkg/util/contextutil/context.go` line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNEi345dPmwiBy5LWb:-NNNbFSR8uCYj0CFTaeu:b8ntoud) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/contextutil/context.go#L95)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

As written, this can return nil even for an expired ctx. End even for a context that was detected to have expired (through calling `Err` on a child produced by `context.WithCancel`. Are you sure that's what you want?
The one caller you have does use `Combine()`, which deals with a nil, but still.
Consider calling `Err()`.
</blockquote></details>

Added `Err()` call.
___
*[`pkg/util/tracing/grpcinterceptor/grpc_interceptor_test.go` line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNIPfx2UpPhsI5RG-9:-NNNPcQQ5uS0EPxK217w:b-xchqbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e24e9c0f39966018deac8f389c4ba3d7b555dcba/pkg/util/tracing/grpcinterceptor/grpc_interceptor_test.go#L134)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think the new tests are worth it. I think all we need to test is that our ctx implementation respects the contract. I think these "integration tests" will no t make much sense to any reader.
</blockquote></details>

Removed these and added an assertion on the error returned by `Err()` to the unit test.

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @otan)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNGwKkBcQuK24kdSm7):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

I set the auth timeout to `1ns` and the returned error generates a stacktrace like this which is what we want:
```
operation "get-user-timeout" timed out after 0s (given timeout 1ns): internal error while retrieving user account: context deadline exceeded
(1) operation "get-user-timeout" timed out after 0s (given timeout 1ns)
Wraps: (2) secondary error attachment
  | captured context error: context deadline exceeded
  | (1) attached stack trace
  |   -- stack trace:
  |   | github.com/cockroachdb/cockroach/pkg/util/contextutil.(*contextWithStacktrace).Err
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:116
  |   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:924
  |   | github.com/cockroachdb/cockroach/pkg/kv.runTxn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:959
  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:922
  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:897
  |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:1471
  |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).DescsTxn
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:1393
  |   | github.com/cockroachdb/cockroach/pkg/sql/sessioninit.(*Cache).GetAuthInfo
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/sessioninit/cache.go:124
  |   | github.com/cockroachdb/cockroach/pkg/sql.retrieveSessionInitInfoWithCache.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:210
  |   | github.com/cockroachdb/cockroach/pkg/sql.retrieveSessionInitInfoWithCache
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:234
  |   | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1.1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:98
  |   | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:131
  |   | github.com/cockroachdb/cockroach/pkg/sql.getUserInfoRunFn.func2
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:190
  |   | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:97
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.authAutoSelectPasswordProtocol.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_methods.go:509
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*AuthBehaviors).Authenticate
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_behaviors.go:56
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).handleAuthentication
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth.go:178
  |   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  |   | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:693
  |   | runtime.goexit
  |   | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1594
  | Wraps: (2) captured context error
  | Wraps: (3) context deadline exceeded
  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) context.deadlineExceededError
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.retrieveSessionInitInfoWithCache
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:237
  | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1.1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:98
  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:131
  | github.com/cockroachdb/cockroach/pkg/sql.getUserInfoRunFn.func2
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:190
  | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo.func1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/user.go:97
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.authAutoSelectPasswordProtocol.func1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_methods.go:509
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*AuthBehaviors).Authenticate
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth_behaviors.go:56
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).handleAuthentication
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth.go:178
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  | 	/Users/ewall/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:693
  | runtime.goexit
  | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1594
Wraps: (4) internal error while retrieving user account
Wraps: (5) context deadline exceeded
  | -- cause hidden behind barrier
  | context deadline exceeded
  | (1) context deadline exceeded
  | Error types: (1) context.deadlineExceededError
Error types: (1) *contextutil.TimeoutError (2) *secondary.withSecondaryError (3) *withstack.withStack (4) *errutil.withPrefix (5) *barriers.barrierErr
```

But you are right, it is possible to wrap `contextWithStacktrace` in another `Context` and call `Err()` on it instead which will not generate a useful stacktrace (I added a test case proving this).
</blockquote></details>

I think the alternative to this would take more time: we need to find each place we get an error from a 3rd party lib and make sure we wrap with `errors.WithStack`. This includes anytime we call `Context.Err()` and return the result.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @otan)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNGwKkBcQuK24kdSm7):*
> I set the auth timeout to 1ns and the returned error generates a stacktrace like this which is what we want:

I think there's two things that make me question whether your example is realistic:
1) By setting such a low timeout, you've made the deadline expiration be detected before the code has a chance to actually do anything - i.e. run an actual RPC in that `InternalDB.DescsTxn()` function. This is not realistic; if the timeout expires while we're in `DescsTxn`, it will expire while we're inside an RPC, and that's when I question whether we're going to get a useful stack trace (or any stack trace at all). In particular, the example that would convince me is when the RPC goes to a remote node - because in the local node case gRPC is bypassed.
2) By setting such a low timeout, even if there's a child `context.WithCancel` on the way (which I claim is common), `Err()` would still be called on our parent `contextWithStack`. This is because of a special case [here](https://github.com/golang/go/blob/cd1fc871568e9947e84377f82c8d7a4882a07067/src/context/context.go#L313). But, in the common case when the `WithCancel` is called before the deadline expires, it wouldn't. This code shows that:
```
func TestXXX(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
	defer cancel()
	ctx = &contextWithStacktrace{
		Context: ctx,
	}
	ctx2, c := context.WithCancel(ctx)
	defer c()
	time.Sleep(100 * time.Millisecond)
	require.NotNil(t, ctx2.Err())
	require.NotNil(t, ctx.(*contextWithStacktrace).ErrWithStacktrace())
}
```
The last line fails; our context has not captured a stack trace.

So I for one would like to see a more realistic example - one where the timeout expires when the task is blocked on something that actually takes some time.
___
*[`pkg/util/contextutil/context_test.go` line 97 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/96483#-NNNoKj-DQuOker0C_Ue:-NNNoKj-DQuOker0C_Uf:bpdkikv) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1ff8c6dd580efe111d109b590453ab3dbe7133/pkg/util/contextutil/context_test.go#L97)):*
> ```Go
> }
> 
> func TestRunWithTimeoutContextWithStacktraceChildContext(t *testing.T) {
> ```

I think we need commentary about what this test shows. 
I also question if there's much point in asserting the limitations of our code.

**Comment by ecwall** [comment]
Making this change instead https://github.com/cockroachdb/cockroach/pull/96659

**Pull request closed without merging**